src/accelerator.rkt

```diff
@@ -82,101 +109,48 @@
-              (and expression** (car expression**)))
-            expression*)
-      [(list operator operands ...) (cons operator (map rewrite operands))]
-      [_ expression*])))
-
-;; TODO: Temporarily copied to avoid cycles. Is there a way to avoid this?
-
-(define (merge-bindings binding1 binding2)
-  (and binding1
-       binding2
-       (let/ec quit
-         (for/fold ([binding binding1]) ([(k v) (in-dict binding2)])
-           (dict-update binding k (λ (x) (if (equal? x v) v (quit #f))) v)))))
-
-(define (pattern-match pattern expr)
-  (match pattern
-   [(? number?)
-    (and (equal? pattern expr) '())]
-   [(? variable?)
-    (list (cons pattern expr))]
-   [(list phead _ ...)
-    (and (list? expr)
-         (equal? (car expr) phead)
-         (= (length expr) (length pattern))
-         (for/fold ([bindings '()])
-             ([pat (cdr pattern)] [subterm (cdr expr)])
-           (merge-bindings bindings (pattern-match pat subterm))))]))
-
-(define (pattern-substitute pattern bindings)
-  ; pattern binding -> expr
-  (match pattern
-   [(? number?) pattern]
-   [(? variable?)
-    (dict-ref bindings pattern)]
-   [(list phead pargs ...)
-    (cons phead (map (curryr pattern-substitute bindings) pargs))]))
-
-(define (rule-apply rule expr)
-  (let ([bindings (pattern-match (rule-input rule) expr)])
-    (if bindings
-        (cons (pattern-substitute (rule-output rule) bindings) bindings)
-        #f)))
-
-(define-accelerator-operator expm1 (real) real (lambda (x) (- (exp x) 1)) ival-expm1)
-(define-accelerator-operator log1p (real) real (lambda (x) (log (+ 1 x))) ival-log1p)
-(define-accelerator-operator hypot (real real) real (lambda (x y) (sqrt (+ (* x x) (* y y)))) ival-hypot)
-(define-accelerator-operator fma (real real real) real (lambda (x y z) (+ (* x y) z)) ival-fma)
-(define-accelerator-operator erfc (real) real (lambda (x) (- 1 (erf x))) ival-erfc)
+                                 #:impl implementation)]))
+
+(define (expand-accelerators expr #:accelerators [ops (all-accelerators)])
+  (define (expand expr)
+    (for/fold ([expr expr]) ([op (in-list ops)])
+      (define info (hash-ref accelerators op))
+      (define-values (lhs rhs) (accelerator-patterns info))
+      (define bindings (pattern-match lhs expr))
+      (if bindings (pattern-substitute rhs bindings) expr)))
+  (let loop ([expr expr])
+    (match (expand expr)
+      [(list 'if cond ift iff) `(if ,(loop cond) ,(loop ift) ,(loop iff))]
+      [(list op args ...) `(,op ,@(map loop args))]
+      [expr expr])))
+
+
+(define-accelerator (expm1 real) real (lambda (x) (- (exp x) 1)))
+(define-accelerator (log1p real) real (lambda (x) (log (+ 1 x))))
+(define-accelerator (hypot real real) real (lambda (x y) (sqrt (+ (* x x) (* y y)))))
+(define-accelerator (fma real real real) real (lambda (x y z) (+ (* x y) z)))
+(define-accelerator (erfc real) real (lambda (x) (- 1 (erf x))))
```

Comment:
This lost the ability to provide interval versions, like `ival-log1p`. This is what's causing the test failures. There are a two possible fixes:
- The most expedient fix is to bring that capability back, it's not hard.
- The longer but eventually better fix is to desugar accelerators on the way into sampling, and then sugar them back inside sampling. This is the long-term plan in sampling.
Luckily we can do both at once, so I say @broughjt you do the first one and I'll work on the second.